### PR TITLE
[lldb][swift] Filter await-resume funclets when setting breakpoints

### PR DESCRIPTION
**Pull request by felipepiovezan**
These funclets only serve to `task_dealloc` previously allocated tasks when returning from an async call, and immediately `task_switch` to the next await-suspend funclet (which contains real user code).

By not filtering out these funclets, any breakpoint on a line with an async call will cause execution to pause 3 times: once before the call, twice when "returning" from the call, which makes for a confusing experience.

The patch does the filtering on `BreakpointResolver::SetSCMatchesByLine`, which is the common code between BreakpointResolverFileLine and BreakpointResolverFileRegex.

We also considered changing the debug line information in any of the many different lowering stages the swift compiler, but this turned out to be very complex to do in a targeted way; more often than not, a handful of early-IR coroutine instructions get expanded into multiple function clones, all inheriting the same debug line information. The current approach also has the advantaged of being easily reversible if we decide to do so.

### CODE REVIEW

**Comment by felipepiovezan** [comment]
I'm very open to other ways of accomplishing this, in case anyone has suggestions

**Comment by jimingham** [comment]
Filtering them in the File & Line resolver seems fine to me.

However, there needs to be a `virtual Language::RemoveInvalidSourceLineMatches` or something like that which does nothing.  Then your function needs to implement that in the Swift Language.  I'm not wedded to the name, but you have to make this generic.  We really want to keep from adding more new #ifdef LLDB_ENABLE_SWIFT in generic files.

**Comment by jimingham** [comment]
You should probably also do the generic stuff on the llvm.org version, cherry-pick that and add your implementation so we don't end up with more diffs on the swift side.

**Comment by felipepiovezan** [comment]
> You should probably also do the generic stuff on the llvm.org version, cherry-pick that and add your implementation so we don't end up with more diffs on the swift side.

I actually had that on my first iteration of this patch, but then realized we would be putting virtual methods with no real implementation upstream, which is generally something that sees push back against (it would basically be dead code, since Swift is not an upstreamed language). If you believe this won't be a problem, I can go the language route (still need to figure out if we have access to a `Language` object from this point)

**Comment by jimingham** [comment]
This is a general feature, where the compiler is emitting some debug information that for its own internal reasons it wishes it could manage not to emit, but can't.  OTOH, the debugger after the fact can.  There's nothing swift specific about that, except currently the swift compiler is the only one that does it.  So I don't think it will be an issue putting in an API here - and that's much preferable to more `#ifdef LLDB_ENABLE_SWIFT`.

The breakpoints offer an option to filter by language, so you should be able to get your hands on the language of the function you are setting the breakpoint in.

**Comment by felipepiovezan** [comment]
Sounds good, I will convert this to a draft while I propose the first commit of this PR upstream

**Comment by felipepiovezan** [comment]
Upstream patch https://github.com/llvm/llvm-project/pull/83908

File: lldb/include/lldb/Target/Language.h
Comment (adrian-prantl) [inline lldb/include/lldb/Target/Language.h]: Doxygen comment?
Also we can probably spell out Ctx?

File: lldb/source/Breakpoint/BreakpointResolver.cpp
Comment (adrian-prantl) [inline lldb/source/Breakpoint/BreakpointResolver.cpp]: Why not just apply the filter right here before entering it into the list?

File: lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: A much better way to implement this would be to walk the demangle tree.
Also this function will incorrectly return true if I had a function that takes an "await resume" function pointer as argument. By just checking the outermost nodes of the demangle tree you can address both problems.

Look at the output of `swift-demangle -expand <mangled name>` and copy any of the functions in TypeSystemSwiftTypeRef.cpp.

Comment (felipepiovezan) [reply lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: > will incorrectly return true if I had a function that takes an "await resume" function pointer as argument

Is this possible? these functions are created by the corosplitter

Comment (adrian-prantl) [reply lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: Probably not, but who knows what future runtime optimizations will be trying to do.

Comment (adrian-prantl) [reply lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: In SwiftLanguageRuntimeNames.cpp there is IsSwiftAsyncFunctionSymbol which you might be able to adapt.

Comment (felipepiovezan) [reply lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: I've looked into those and they seem like a nicer way to do these checks indeed.
That said, there are at least two bugs in their implementation that we'll need to fix before being able to reuse them.

**Comment by felipepiovezan** [comment]
I've re-implemented the swift side of things with the suggestions and marked this patch as ready for review.

Note that the first commit here is identical to https://github.com/llvm/llvm-project/pull/83908/files, and I'll convert it to an actual cherry-pick once it gets merged upstream (won't merge this PR until that's done)

File: lldb/include/lldb/Target/Language.h
Comment (adrian-prantl) [inline lldb/include/lldb/Target/Language.h]: ///

File: lldb/include/lldb/Target/Language.h
Comment (adrian-prantl) [inline lldb/include/lldb/Target/Language.h]: how about `IgnoreForLineBreakpoints` ?

File: lldb/source/Breakpoint/BreakpointResolver.cpp
Comment (adrian-prantl) [inline lldb/source/Breakpoint/BreakpointResolver.cpp]: can this be range-based?

File: lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp
Comment (adrian-prantl) [inline lldb/source/Plugins/Language/Swift/SwiftLanguage.cpp]: Great!

File: lldb/source/Breakpoint/BreakpointResolver.cpp
Comment (felipepiovezan) [reply lldb/source/Breakpoint/BreakpointResolver.cpp]: It can, I just didn't want to add to the diff. As much as github is saying this is new code, it's not new code...

**Comment by felipepiovezan** [comment]
@adrian-prantl I've addressed your comments on the upstream pr https://github.com/llvm/llvm-project/pull/83908

File: lldb/include/lldb/Target/Language.h
Comment (adrian-prantl) [inline lldb/include/lldb/Target/Language.h:359]: Side note: let's try to upstream this abstract interface later.

**Review by adrian-prantl — APPROVED** [review]

**Comment by felipepiovezan** [comment]
Now with the official cherry-pick of the upstream lldb commit.

**Comment by felipepiovezan** [comment]
@swift-ci test

**Review by adrian-prantl — APPROVED** [review]

File: lldb/source/Target/TargetProperties.td
Comment (kastiglione) [inline lldb/source/Target/TargetProperties.td:365]: Is there a naming convention around boolean properties? Should this be prefixed with enable-, or not? It seems possibly redundant to me. Maybe the name should be `filter-for-line-breakpoints` and the values `enable` and `disable`? I'm not strongly opposed to the name, but it did strike me as a little unconventional.

Comment (felipepiovezan) [reply lldb/source/Target/TargetProperties.td:365]: Current practice has all "enable"-like properties using boolean values, and most of them mentioning "enable" in the actual setting flag, though there are exceptions to this second point

**Comment by felipepiovezan** [comment]
Superseded by https://github.com/swiftlang/llvm-project/pull/9112

**Pull request closed without merging**
